File: test/VisualBasic/CodeCracker.Test/Refactoring/AllowMembersOrderingAnalyzerTests.vb
Imports Microsoft.CodeAnalysis.Diagnostics
Imports CodeCracker.VisualBasic.Refactoring
Imports Xunit
Imports Microsoft.CodeAnalysis.Testing

Namespace Refactoring
    Public Class AllowMembersOrderingAnalyzerTests
        Inherits CodeFixVerifier

        Protected Overrides Function GetDiagnosticAnalyzer() As DiagnosticAnalyzer
            Return New AllowMembersOrderingAnalyzer()
        End Function

        <Theory>
        <InlineData("Class")>
        <InlineData("Structure")>
        Public Async Function AllowMembersOrderingForEmptyTypeShouldNotTriggerDiagnostic(typeDeclaration As String) As Task
            Dim test = String.Format("{0} Foo
End {0}", typeDeclaration)
            Await VerifyBasicHasNoDiagnosticsAsync(test)
        End Function

        <Theory>
        <InlineData("Class")>
        <InlineData("Structure")>
        Public Async Function AllowMembersOrderingForOneMemberShouldNotTriggerDiagnostic(typeDeclaration As String) As Task
            Dim test = String.Format("{0} Foo
    Function bar()
        Return 0
    End Function
End {0}", typeDeclaration)
            Await VerifyBasicHasNoDiagnosticsAsync(test)
        End Function

        <Theory>
        <InlineData("Class")>
        <InlineData("Structure")>
        Public Async Function AllowMembersOrderingForMoreThanOneMembersHouldTriggerDiagnostic(typeDeclaration As String) As Task
            Dim test = String.Format("{0} Foo
    Function bar()
        Return 0
    End Function
    Sub car()
    End Sub
End {0}", typeDeclaration)

            Dim expected = New DiagnosticResult(AllowMembersOrderingAnalyzer.Id, Microsoft.CodeAnalysis.DiagnosticSeverity.Hidden) _
                .WithLocation(2, 14 + typeDeclaration.Length) _
                .WithMessage(AllowMembersOrderingAnalyzer.MessageFormat)
            Await VerifyBasicDiagnosticAsync(test, expected)
        End Function
    End Class
End Namespace

File: test/CSharp/CodeCracker.Test/Refactoring/AllowMembersOrderingAnalyzerTests.cs
using CodeCracker.CSharp.Refactoring;
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.Diagnostics;
using Microsoft.CodeAnalysis.Testing;
using Xunit;
namespace CodeCracker.Test.CSharp.Refactoring
{
    public class AllowMembersOrderingAnalyzerTests : CodeFixVerifier
    {
        protected override DiagnosticAnalyzer GetDiagnosticAnalyzer()
        {
            return new AllowMembersOrderingAnalyzer();
        }

        [Theory]
        [InlineData("class")]
        [InlineData("struct")]
        public async void AllowMembersOrderingForEmptyTypeShouldNotTriggerDiagnostic(string typeDeclaration)
        {
            var test = @"
            " + typeDeclaration + @" Foo
            {
            }";

            await VerifyCSharpHasNoDiagnosticsAsync(test);
        }

        [Theory]
        [InlineData("class")]
        [InlineData("struct")]
        public async void AllowMembersOrderingForOneMemberShouldNotTriggerDiagnostic(string typeDeclaration)
        {
            var test = @"
            " + typeDeclaration + @" Foo
            {
                int bar() { return 0; }
            }";

            await VerifyCSharpHasNoDiagnosticsAsync(test);
        }

        [Theory]
        [InlineData("class")]
        [InlineData("struct")]
        public async void AllowMembersOrderingForMoreThanOneMemberShouldTriggerDiagnostic(string typeDeclaration)
        {
            var test = @"
            " + typeDeclaration + @" Foo
            {
                int bar() { return 0; }
                void car() { }
            }";

            var expected = new DiagnosticResult(DiagnosticId.AllowMembersOrdering.ToDiagnosticId(), DiagnosticSeverity.Hidden)
                .WithLocation(2, 14 + typeDeclaration.Length)
                .WithMessage(AllowMembersOrderingAnalyzer.MessageFormat);

            await VerifyCSharpDiagnosticAsync(test, expected);
        }
    }
}